### PR TITLE
Allow the user to override the timeout of the xcode list command, also shows the current issues with the implementation (fastlane/gym#143)

### PR DESCRIPTION
**Pull request by lacostej**
See fastlane/gym#143

### CODE REVIEW

File: lib/fastlane_core/project.rb [hunk 1]
```diff
@@ -275,4 +274,21 @@ def parsed_info
       @parsed_info
     end
   end
+
+  # @internal to module
+  def self.xcode_list_timeout
+    timeout = ENV['FASTLANE_XCODE_LIST_TIMEOUT']
+    timeout = timeout.to_i if timeout
+    timeout = 10 unless timeout
```
Comment (KrauseFx) [inline lib/fastlane_core/project.rb]:
```ruby
return (ENV['FASTLANE_XCODE_LIST_TIMEOUT'] || 10).to_i
```

**Comment by KrauseFx** [comment]
Is there any good way to test this PR?

**Comment by lacostej** [comment]
@KrauseFx I mentioned one way of testing the PR in https://github.com/fastlane/gym/issues/143#issuecomment-179770584

(setup Gemfile pointing to the PR + env variable)

Note that the PR doesn't change the default behaviour if you do not set the environment variable and that it adds tests for the different scenarios (timeout vs not timeout). We're missing tests for the command_executor#execute method but this is mostly taken care of by #73

**Comment by thelvis4** [comment]
Is there anything to be done before merging this PR?
It would be nice to have it ASAP.

P.S. I tested it and it worked well for me.

File: lib/fastlane_core/project.rb [hunk 2]
```diff
@@ -275,4 +276,18 @@ def parsed_info
...
+  # @internal to module
+  def self.xcode_list_timeout
+    (ENV['FASTLANE_XCODE_LIST_TIMEOUT'] || 10).to_i
+  end
+
+  # runs the specified command and kills it if timeouts
+  # @raises Timeout::Error if timeout is passed
+  # @returns the output
+  # FIXME: this is not working exactly as it should. gym#143
+  def self.run_command(command, timeout: 0)
```
Comment (KrauseFx) [inline lib/fastlane_core/project.rb]: This method shouldn't be added to the `FastlaneCore` scope.

File: lib/fastlane_core/project.rb [hunk 3]
```diff
@@ -275,4 +276,18 @@ def parsed_info
...
+
+  # @internal to module
+  def self.xcode_list_timeout
+    (ENV['FASTLANE_XCODE_LIST_TIMEOUT'] || 10).to_i
+  end
+
+  # runs the specified command and kills it if timeouts
+  # @raises Timeout::Error if timeout is passed
+  # @returns the output
+  # FIXME: this is not working exactly as it should. gym#143
```
Comment (KrauseFx) [inline lib/fastlane_core/project.rb]: We shouldn't have FIXMEs in the code

File: spec/project_spec.rb [hunk 4]
```diff
@@ -117,4 +117,61 @@
...
+      it "runs more complicated commands" do
+        cmd = "ruby -e 'sleep 0.1; puts \"HI\"'"
+        expect(FastlaneCore.run_command(cmd)).to eq("HI\n")
+      end
+
+      it "should timeouts and kills" do
+        text = "FOOBAR" # random text
+        count = count_processes(text)
+        cmd = "ruby -e 'sleep 3; puts \"#{text}\"'"
+        # this doesn't work ?!
```
Comment (KrauseFx) [inline spec/project_spec.rb]: Hu?

File: lib/fastlane_core/project.rb (hunk 3 above)
Comment (lacostej) [inline lib/fastlane_core/project.rb]: The current implementation (and the new one) contain an issue. This was just a reminder of that problem. I can remove the FIXME if you want. The issue is demonstrated in the unit test were you added your "hu?" comment. I can open a new issue in github instead.

File: lib/fastlane_core/project.rb (hunk 2 above)
Comment (lacostej) [inline lib/fastlane_core/project.rb]: I described the other method as "internal to the module". This one was meant to as well.

I can move it out of the FastlaneCore scope, but I still want to make it somewhat 'private'

Is there any guideline on how to properly implement stuff that one wants clearly not exposed (not used by users of the spaceship gem) yet still available for unit testing ?

File: spec/project_spec.rb (hunk 4 above)
Comment (lacostej) [inline spec/project_spec.rb]: I've seen that issue multiple times in the past weeks. E.g. https://github.com/lacostej/spaceship/blob/client_hardening/spec/client_spec.rb#L41-L43

Not sure why this is like that... Do you have any idea ?

**Comment by revile** [comment]
Ditto. Been waiting for this to get fixed, we have randomly failing builds all the time (we use a couple of old mac minis as build servers)

**Comment by jk** [comment]
@KrauseFx Sorry, I don't want to bother, but the issue causes a significant amout of trouble for us. Any ETA or is worth to remove gym from our build chain to workaround this issue?

**Comment by asfalcone** [comment]
This looks good to me :watch:

**Comment by KrauseFx** [comment]
Thanks everyone, bundling a new release now :+1:

**Comment by lacostej** [comment]
@hufkens  try setting up the FASTLANE_XCODE_LIST_TIMEOUT environment variable, either to a number of seconds (e.g. 60) or 0 if you don't want it to timeout at all (and potentially block forever). Please report if that helped you or not.

Note: if you are affected, fastlane should fail with an error message telling you more or less the above. Didn't you see it ?

**Comment by hufkens** [comment]
Looks like that works. Thanks

**Comment by lacostej** [comment]
@hufkens great! 

could you any any chance confirm or not if the error message popped up when you didn't use the environment variable. This is the only place where the feature is documented right now :)

**Comment by hufkens** [comment]
It worked the first time, but I keep getting this error:
```[08:44:16]: xcodebuild -list timed-out. You might need to recreate the user schemes. See https://github.com/fastlane/gym/issues/143```
So no. It didn't work and I didn't see the error message.

**Comment by lacostej** [comment]
@hufkens Check the version of fastlane_core you are using.

Your message was in 0.36.5
https://github.com/fastlane/fastlane_core/blob/0.36.5/lib/fastlane_core/project.rb#L261

The fix came in 0.36.6
https://github.com/fastlane/fastlane_core/blob/0.36.6/lib/fastlane_core/project.rb#L261

**Comment by hufkens** [comment]
Thanks for the quick follow up. I always update fastlane, but that said everything was up to date. But gem update fastlane_core updated to version 0.36.6 and now the build seems to be running.

Fingers crossed that the ENV variable trick works now and thanks again.

**Pull request merged**